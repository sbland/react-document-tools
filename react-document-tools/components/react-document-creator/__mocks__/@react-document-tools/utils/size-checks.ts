export const getPixelRatio = () => {
  return 1;
};

// export const calculateCellHeight = (text, width) => {
//   // eslint-disable-next-line no-unused-vars
//   const [lineCount, finalRemainingWidth] = (text || '').split(' ').reduce(
//     (acc, word) => {
//       const [currLineCount, remainingWidth] = acc;
//       const wordWidth = word.length * charWidth;
//       const startNewLine = remainingWidth < width && wordWidth > remainingWidth;
//       const newRemainingWidth = startNewLine ? width - wordWidth : remainingWidth - wordWidth;
//       const newLineCount = startNewLine ? currLineCount + 1 : currLineCount;
//       return [newLineCount, newRemainingWidth];
//     },
//     [1, width]
//   );
//   return lineHeight * lineCount;
// };

function parseFont(font) {
  let fontFamily = null;

  let fontSize = null;
  let fontStyle = 'normal';
  let fontWeight = 'normal';
  let fontVariant = 'normal';
  let lineHeight = 'normal';

  const elements = font.split(/\s+/);
  elements.forEach((element) => {
    switch (element) {
      case 'normal':
        break;

      case 'italic':
      case 'oblique':
        fontStyle = element;
        break;

      case 'small-caps':
        fontVariant = element;
        break;

      case 'bold':
      case 'bolder':
      case 'lighter':
      case '100':
      case '200':
      case '300':
      case '400':
      case '500':
      case '600':
      case '700':
      case '800':
      case '900':
        fontWeight = element;
        break;

      default:
        if (!fontSize) {
          const parts = element.split('/');
          [fontSize] = parts;
          if (parts.length > 1) [, lineHeight] = parts;
          break;
        }

        fontFamily = element;
        // eslint-disable-next-line prefer-template
        if (elements.length) fontFamily += ' ' + elements.join(' ');
      // break outer;
    }
  });

  return {
    fontStyle,
    fontVariant,
    fontWeight,
    fontSize,
    lineHeight,
    fontFamily,
  };
}

export const getTextAreaSize = (txt, font) => {
  const parsedFont = font ? parseFont(font) : {};
  const { fontSize, lineHeight } = parsedFont as any;
  const fontSizeParsed = fontSize ? parseFloat(fontSize.split('mm')[0]) : 99;
  const lineHeightParsed = lineHeight
    ? parseFloat(lineHeight.split('mm')[0])
    : 99;
  return {
    width: txt ? fontSizeParsed * txt.length : 0,
    height: lineHeightParsed,
  };
};

export const getTextWidth = (txt, font) => {
  const parsedFont = font ? parseFont(font) : {};
  const { fontSize } = parsedFont as any;
  const fontSizeParsed = fontSize ? parseFloat(fontSize.split('mm')[0]) : 99;
  return txt ? fontSizeParsed * txt.length : 0;
};

export const getTextHeight = (txt, font) => {
  /* https://stackoverflow.com/a/31305410/10259813 */

  const parsedFont = font ? parseFont(font) : {};
  const { lineHeight } = parsedFont as any;
  const lineHeightParsed = lineHeight
    ? parseFloat(lineHeight.split('mm')[0])
    : 99;
  return lineHeightParsed;
};

export const calculateBodyDimensions = (settings) => {
  const {
    paperHeight,
    paperWidth,
    headerHeight,
    footerHeight,
    marginTop,
    marginBottom,
    marginLeft,
    marginRight,
  } = settings;
  const height =
    paperHeight - marginTop - marginBottom - headerHeight - footerHeight;

  const width = paperWidth - marginLeft - marginRight;
  const pageHeight = paperHeight - marginTop - marginBottom;
  return { width, height, pageHeight };
};

export const mmToPx = (mm) => {
  const pixelRatio = getPixelRatio();
  return mm * pixelRatio;
};
