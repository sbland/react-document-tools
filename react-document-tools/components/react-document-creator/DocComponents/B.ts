import { Uid } from '@react_db_client/constants.client-types';
import { EContentType, ISettings, ITemplateContentItem } from '../lib';
import { getNextId } from '../utils/idGenerator';

export interface IBItem extends ITemplateContentItem {}

export interface BArgs {
  align?: 'left' | 'center' | 'right';
  font?: string;
  style?: React.CSSProperties;
}

export const B =
  (text: string, { align = 'left', font = null, style = {} }: BArgs = {}, id?: Uid) =>
  (settings: ISettings): IBItem => ({
    id,
    uid: getNextId(),
    type: EContentType.B,
    text,
    height: settings.lineHeight,
    align,
    style: { ...style, ...(settings?.styleOverrides?.b || {}) },
    font: font || settings.defaultFont,
  });
