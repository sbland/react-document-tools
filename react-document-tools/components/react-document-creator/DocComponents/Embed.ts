import { EContentType, ISettings, ITemplateContentItem } from '../lib';
import { getNextId } from '../utils/idGenerator';

export enum EEmbedType {
  PDF = 'PDF',
}

export interface IEmbedItem extends ITemplateContentItem {
  src: string;
  embedType: EEmbedType;
}

export interface IEmbedArgs {
  align?: 'left' | 'center' | 'right';
  style?: React.CSSProperties;
}

/**
 * Embed different files.
 * @param {String} src path to src for embed
 * @param {string} type one of ['PDF']
 * @param {number} height
 * @param {number} width
 */
export const Embed =
  (
    id: string,
    src: string,
    type: EEmbedType,
    height: 'auto' | number,
    width: 'auto' | number,
    { align = 'left', style = {} }: IEmbedArgs = {}
  ) =>
  (settings: ISettings): IEmbedItem => ({
    id,
    uid: getNextId(),
    type: EContentType.EMBED,
    embedType: type,
    src,
    height,
    width,
    align,
    style: { ...style, ...(settings?.styleOverrides?.div || {}) },
  });
