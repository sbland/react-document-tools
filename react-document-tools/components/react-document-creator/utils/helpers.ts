import { AnyTemplateComponentType, P } from '../DocComponents';
import {
  ElementTypes,
  ISettings,
  // AllElementTypes,
  ITemplate,
  ITemplateProcessed,
  // AllElementTypesProcessed,
} from '../lib';

export const processElement = (
  element: ElementTypes,
  settings: ISettings
): AnyTemplateComponentType => {
  if (!settings) throw Error('Settings is undefined');
  if (typeof element === 'function') return element(settings);
  if (typeof element === 'string') return P(element)(settings);
  if (Array.isArray(element)) {
    throw Error('processElement: Array is not supported.');
  }
  return element;
};

export const processContent = (
  content: ElementTypes[],
  settings: ISettings
): AnyTemplateComponentType[] => {
  return content?.map((c) => processElement(c, settings));
};

export const processBody = (body, settings) => ({
  ...body,
  content: processContent(body.content, settings),
});

export const processTemplate = (template: ITemplate, settings: ISettings): ITemplateProcessed => {
  const header =
    template.header &&
    template.header({
      date: 'unknown',
      pageNo: -1,
      pageCount: -1,
    });
  const footer =
    template.footer &&
    template.footer({
      date: 'unknown',
      pageNo: -1,
      pageCount: -1,
    });
  return {
    header: header
      ? {
          ...header,
          content: processContent(header.content, settings),
        }
      : null,
    footer: footer
      ? {
          ...footer,
          content: processContent(footer.content, settings),
        }
      : null,
    body: template.body
      ? {
          ...template.body,
          content: processContent(template.body.content, settings),
        }
      : null,
  };
};

export const getFontString = (
  weight: number,
  textHeight: number,
  lineHeight: number,
  fontFamily: string
) => `${weight} ${textHeight.toFixed(0)}mm/${lineHeight}mm  ${fontFamily}`;

export const calcLineHeight = (textHeight: number) => textHeight * (5 / 3.8);
