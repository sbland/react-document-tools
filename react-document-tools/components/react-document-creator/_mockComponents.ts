/* Mock components for testing */

import { EFilterType } from '@react_db_client/constants.client-types';
import { styleDefaults } from './ComponentStyles';
import { DEFAULT_SETTINGS } from './constants';
import {
  H1,
  H,
  Section,
  Span,
  P,
  Div,
  Img,
  Table,
  PageBreak,
  Hr,
  TableCell,
  B,
  Container,
  Embed,
  FullPage,
} from './DocComponents';
import { EEmbedType } from './DocComponents/Embed';
import { EContentType, ISettings, ITemplate } from './lib';

export const MOCK_DIV = () => Div('Hello World', 100, 20);
export const MOCK_DIV_WITH_CHILDREN = () => Div(P('Hello World'), 100, 20);
export const MOCK_DIV_CENTER = () =>
  Div('Hello World', 100, 20, { align: 'center' });
export const MOCK_CONTAINER = () => Container('Hello World', 100, 20);
export const MOCK_HR = () => Hr(10, 20);
export const MOCK_HEADER = () => H('demo header', 3);
export const MOCK_SPAN = () => Span('demo span');
export const MOCK_PARAGRAPH = () => P('demo paragraph');
export const MOCK_BOLD = () => B('demo bold text');
export const MOCK_PAGE_BREAK = () => PageBreak();

export const MOCK_IMAGE_EXAMPLE = () =>
  Img('/images/logo.png', 50, 20, 'Demo Img');
export const MOCK_TABLE_EXAMPLE = () =>
  Table({
    uid: -99,
    id: 'Demo Table',
    columnWidths: [20, 30, 30, 60],
    rowHeights: [20, 30],
    content: [
      ['Hello', 'World', 'This', 'Is some information'],
      ['Another', 'Place', 'This', 'Is some information'],
    ],
    headings: ['NAME', 'PLACE', 'INFO', 'DESCRIPTION'],
  });

export const MOCK_TABLE_BIG_EXAMPLE = () =>
  Table({
    uid: -98,
    id: 'Demo Table Big',
    columnWidths: [20, 30, 30, 60],
    rowHeights: [20, 30, 90, 60, 10],
    content: [
      ['Hello', 'World', 'This', 'Is some information'],
      ['Another', 'World', 'This', 'Is some information'],
      ['Another', 'Place', 'This', 'Is some information'],
      ['Another', 'Place', 'This', 'Is some information'],
      ['Another', 'Land', 'This', 'Is some information'],
    ],
    headings: ['NAME', 'PLACE', 'INFO', 'DESCRIPTION'],
    headingsMargin: 10,
  });

export const longText = [
  ' This is some really long text',
  'and more text that should overflow',
].join(' ');

export const longTextB =
  'Consort CBH102 102x76x3mm Grade 304 SSS Finish Ball Bearing Butt Hinge, Load Bearing Capacity Up To 120kg and Durability Tested To Over 200,000 Cycles. Tested to BS EN 1935:2002, Grade 13 (Classification Code 4.7.6.1.1.4.0.13). Fire Door Approved: Suitable for use on up to FD120 Timber and FD240 Metal Door and Frame Applications. Certifire Warrington Certificate of Approval No. CF5511';

const items = [{ a: 'a', b: 'b', c: 10, d: { filePath: 'd.jpg' } }];

const headings = [
  { type: EFilterType.text, uid: 'a', width: 30 },
  { type: EFilterType.textLong, uid: 'b', width: 30 },
  { type: EFilterType.number, uid: 'c', width: 30 },
  { type: EFilterType.image, uid: 'd', width: 50, height: 30 },
];
export const MOCK_TABLE_CALCULATE_EXAMPLE = () =>
  Table({
    // 'calculate',
    uid: 97,
    id: 'Calulate Table',
    columnWidths: [20, 40, 40, 30],
    rowHeights: 'calculate',
    content: [
      ['Hello', 'World', Div(longText), Div('Is some information')],
      ['Another', 'World', 'This', Div('boxed content', 50)],
      ...items.map((item) =>
        headings.map((h) =>
          TableCell(item[h.uid], h.width, h.height || 'auto', h.type, 3)
        )
      ),
    ],
    headings: ['NAME', 'PLACE', 'INFO', 'DESCRIPTION'],
    cellPadding: 3,
  });

export const MOCK_TABLE_COMPLEX_EXAMPLE = () =>
  Table({
    id: 'Complex table',
    uid: 96,
    columnWidths: [20, 30, 30, 60],
    rowHeights: [20, 30],
    content: [
      ['Hello', 'World', 'This', MOCK_IMAGE_EXAMPLE()],
      ['Hello', 'World', 'This', MOCK_IMAGE_EXAMPLE()],
    ],
    headings: ['NAME', 'PLACE', 'INFO', 'DESCRIPTION'],
    headingsHeight: 30,
    cellPadding: 3,
  });

  const multiLineText = `Row 1
  Row 2

  After break
  More text

  \n\n\n\n

  `;


const longMultiLineText = `Row 1
Row 2

After break
More text

\n\n\n\nMore text

\n\n\n\nMore text

\n\n\n\nMore text

\n\n\n\nMore text

\n\n\n\nMore text

\n\n\n\nMore text

\n\n\n\n

`;


export const MOCK_JSON_PAGE = (): ITemplate => ({
  header: ({ date, pageNo, pageCount }) => ({
    content: [H1('Header'), P(date), P(String(pageNo)), P(String(pageCount))],
    options: {},
  }),
  body: {
    content: [
      H1('Head'),
      P('Hello'),
      Div(multiLineText, "calculate", 100),
      Div(longMultiLineText, "calculate", 100),
      Section([P('child 1')]),
      MOCK_IMAGE_EXAMPLE(),
      MOCK_TABLE_EXAMPLE(),
      MOCK_TABLE_CALCULATE_EXAMPLE(),
    ],
    options: {},
  },
  footer: ({ date, pageNo, pageCount }) => ({
    content: [H1('Footer'), P(date), P(String(pageNo)), P(String(pageCount))],
    options: {},
  }),
  type: EContentType.SECTION,
});

export const MOCK_EMBED = () =>
  Embed('demo_embed', 'demofile.pdf', EEmbedType.PDF, 100, 200);
export const MOCK_FULL_PAGE = () => FullPage([P('Hello')], 'example mock page');

export const MOCK_OPTIONS: ISettings = {
  ...DEFAULT_SETTINGS,
  paperHeight: 297,
  paperWidth: 210,
  headerHeight: 30,
  footerHeight: 30,
  marginTop: 10,
  marginBottom: 10,
  marginLeft: 10,
  marginRight: 10,
  styleOverrides: styleDefaults(
    DEFAULT_SETTINGS.fontWeight,
    DEFAULT_SETTINGS.textHeight,
    DEFAULT_SETTINGS.lineHeight,
    DEFAULT_SETTINGS.fontFamily,
    DEFAULT_SETTINGS.headingHeight
  ),
};

export const MOCK_BODY_HEIGHT =
  DEFAULT_SETTINGS.internalHeight -
  DEFAULT_SETTINGS.headerHeight -
  DEFAULT_SETTINGS.footerHeight;
export const MOCK_PAGE_HEIGHT = DEFAULT_SETTINGS.paperHeight;
