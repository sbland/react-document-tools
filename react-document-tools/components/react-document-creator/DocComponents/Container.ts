import { EContentType, ElementTypes, ISettings, ITemplateContentItem } from '../lib';
import { processElement } from '../utils/helpers';
import { getNextId } from '../utils/idGenerator';

export interface IContainerItem extends ITemplateContentItem {
  type: EContentType.CONTAINER;
}

// The container type is used to place any content in a div with specified height and width.
// It is the responsability of the user to ensure this content is safe.
export const Container =
  (content: ElementTypes, height: number, width: number, { id = 'MISSING_ID', style = {} } = {}) =>
  (settings: ISettings): IContainerItem => ({
    id,
    uid: getNextId(),
    type: EContentType.CONTAINER,
    content: processElement(content, settings),
    height,
    width,
    style: { ...style, ...(settings?.styleOverrides?.div || {}) },
  });
