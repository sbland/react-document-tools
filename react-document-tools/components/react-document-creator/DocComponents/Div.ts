import { EContentType, ElementTypes, ISettings, ITemplateContentItem } from '../lib';
import { calculateCellHeightMultiType } from '../utils/calculateCellSize';
import { processElement } from '../utils/helpers';
import { getNextId } from '../utils/idGenerator';

export const processDivContent = (
  content: ElementTypes[] | ElementTypes | string,
  width,
  settings
) => {
  if (!content) return '';
  if (typeof content === 'string') {
    if (!content.includes('\n')) return content;
    return content.split('\n').map((t) => {
      /* eslint-ignore-next-line no-use-before-define */
      return Div(t, 'calculate', width)(settings);
    });
  }
  if (Array.isArray(content)) return content.map((c) => processDivContent(c, width, settings));
  return processElement(content, settings);
};

export interface IDivItem extends ITemplateContentItem {}

export interface DivArgs {
  align?: 'left' | 'center' | 'right';
  font?: string;
  style?: React.CSSProperties;
  padding?: number;
  id?: string;
}

export const Div =
  (
    content: ElementTypes | ElementTypes[] | string,
    height: number | 'auto' | 'calculate' = 'calculate',
    width: number | 'auto' = 'auto',
    { align = 'left', font = null, style = {}, padding = 0, id = 'MISSING_ID' }: DivArgs = {}
  ) =>
  (settings: ISettings): IDivItem => {
    const fontUsed = font || settings.defaultFont;
    const contentProcessed = processDivContent(content, width, settings);
    const heightCalculated =
      height === 'calculate'
        ? Math.max(
            settings.lineHeight,
            width === 'auto'
              ? 0
              : calculateCellHeightMultiType(contentProcessed, width, padding, fontUsed)
          )
        : height;
    return {
      id,
      uid: getNextId(),
      type: EContentType.DIV,
      content: contentProcessed,
      height: heightCalculated,
      width,
      align,
      style: { ...(settings?.styleOverrides?.div || {}), ...style },
      font: fontUsed,
    };
  };
