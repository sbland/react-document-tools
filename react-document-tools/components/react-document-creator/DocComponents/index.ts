import { IBItem } from './B';
import { IDivItem } from './Div';
import { IContainerItem } from './Container';
import { IEmbedItem } from './Embed';
import { IFullPageItem } from './FullPage';
import { IHItem } from './H';
import { IHrItem } from './Hr';
import { IImgItem } from './Img';
import { IPItem } from './P';
import { IPageBreakItem } from './PageBreak';
import { ISectionItem } from './Section';
import { ISpanItem } from './Span';
import { ITableItem } from './Table';

export { B } from './B';
export { Div } from './Div';
export { Container } from './Container';
export { Embed } from './Embed';
export { FullPage } from './FullPage';
export { H, H1 } from './H';
export { Hr } from './Hr';
export { Img } from './Img';
export { P } from './P';
export { PageBreak } from './PageBreak';
export { Section } from './Section';
export { Span } from './Span';
export { Table, TableCell } from './Table';

export type { IBItem } from './B';
export type { IDivItem } from './Div';
export type { IContainerItem } from './Container';
export type { IEmbedItem } from './Embed';
export type { IFullPageItem } from './FullPage';
export type { IHItem } from './H';
export type { IHrItem } from './Hr';
export type { IImgItem } from './Img';
export type { IPItem } from './P';
export type { IPageBreakItem } from './PageBreak';
export type { ISectionItem } from './Section';
export type { ISpanItem } from './Span';
export type { ITableItem } from './Table';

export type AnyTemplateComponentType =
  | IBItem
  | IDivItem
  | IContainerItem
  | IEmbedItem
  | IFullPageItem
  | IHItem
  | IHrItem
  | IImgItem
  | IPItem
  | IPageBreakItem
  | ISectionItem
  | ISpanItem
  | ITableItem;
