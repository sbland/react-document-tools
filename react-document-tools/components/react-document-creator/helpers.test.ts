import {
  defaultFont,
  DEFAULT_SETTINGS,
  lineHeight,
  maxRowHeight,
  minRowHeight,
} from './constants';
import { IBItem, IDivItem, IPItem, P, TableCell } from './DocComponents';
import {
  calculateCellHeight,
  calculateCellHeightMultiType,
  calculateTableColumnWidths,
  calculateTableRowHeight,
  calculateTableRowHeights,
} from './utils/calculateCellSize';
import { processElement } from './utils/helpers';
import { longText, longTextB } from './_mockComponents';
import { EContentType } from './lib';

jest.mock('@react-document-tools/utils.size-checks');

describe('Calculate Row Heights', () => {
  describe('Calculate Table Row Heights', () => {
    const font = defaultFont;
    test('should use minimum row height', () => {
      const content = [['a', 'b', 'c']];
      const rowHeights = calculateTableRowHeights(
        content,
        [10, 10, 10],
        0,
        font,
        DEFAULT_SETTINGS
      );
      expect(rowHeights).toEqual([minRowHeight]);
    });
    test('should use maximum row height', () => {
      const content = [
        ['a', 'b', `${longText}${longText}${longText}${longText}${longText}`],
      ];
      const rowHeights = calculateTableRowHeights(
        content,
        [60, 60, 10],
        0,
        font,
        DEFAULT_SETTINGS
      );
      expect(rowHeights).toEqual([maxRowHeight]);
    });
    test('should Calculate the row height based on the height of the tallest cell', () => {
      const content = ['a', 'b', longText];
      const rowHeight = calculateTableRowHeight(
        content,
        [20, 20, 20],
        0,
        font,
        DEFAULT_SETTINGS
      );
      const expectedLineCount = 12;
      expect(rowHeight).toEqual(lineHeight * expectedLineCount);
    });
    test('should implement padding in height calculation', () => {
      const content = ['a', 'b', longText];
      const padding = 3;
      const rowHeight = calculateTableRowHeight(
        content,
        [20, 20, 20],
        padding,
        font,
        DEFAULT_SETTINGS
      );
      const expectedLineCount = 13;
      expect(rowHeight).toEqual(lineHeight * expectedLineCount + padding);
    });
    test('should use height field if supplied', () => {
      const content = ['a', 'b', { text: 'abc', height: 90 }];
      const rowHeight = calculateTableRowHeight(
        content,
        [10, 10, 10],
        0,
        font,
        DEFAULT_SETTINGS
      );
      expect(rowHeight).toEqual(90);
    });

    test('should Calculate the row height based on the height of the tallest cell real world example', () => {
      const content = ['a', 'b', longText];
      const rowHeight = calculateTableRowHeight(
        content,
        [20, 20, 58],
        0,
        font,
        DEFAULT_SETTINGS
      );
      expect(rowHeight).toBeGreaterThan(lineHeight * 1);
    });
  });

  describe('Calculate Cell Height Multi', () => {
    test('should calculate cell height', () => {
      const cellData: IDivItem = {
        id: 'DEMOID',
        uid: 'DEMOID',
        type: EContentType.DIV,
        content: [
          {
            uid: 'DEMOID',
            type: EContentType.B,
            text: 'Value',
            height: 4.5,
            align: 'left',
            style: {
              padding: 0,
              margin: 0,
              whiteSpace: 'nowrap',
              overflow: 'hidden',
              textOverflow: 'ellipsis',
              fontWeight: 'bold',
            },
            font: '100 4mm/4.5mm  Arial',
          } as IBItem,
          {
            uid: 'DEMOID',
            type: EContentType.P,
            text: '(abc)',
            height: 4.5,
            align: 'left',
            style: {
              padding: 0,
              margin: 0,
              whiteSpace: 'nowrap',
              overflow: 'hidden',
              textOverflow: 'ellipsis',
            },
            font: '100 4mm/4.5mm  Arial',
          } as IPItem,
        ],
        height: 'auto',
        width: 'auto',
        align: 'left',
        style: {
          padding: 0,
          margin: 0,
          overflow: 'hidden',
          textOverflow: 'ellipsis',
        },
        font: '100 4mm/4.5mm  Arial',
      };
      const columnWidth = 95;
      const cellPadding = 4;
      const font = '5mm/4.56mm Arial';
      const expectedHeight = 4.5 + 4.5 + cellPadding;
      const calculatedHeight = calculateCellHeightMultiType(
        cellData,
        columnWidth,
        cellPadding,
        font
      );
      expect(calculatedHeight).toEqual(expectedHeight);
    });
  });
  describe('Calculate Cell Height', () => {
    const font = defaultFont;
    test('single row should be half size of 2 rows', () => {
      const width = 25;
      const text = 'hello';
      const height = calculateCellHeight(text, width, font);

      const textB = 'hello world';
      const heightB = calculateCellHeight(textB, width, font);
      expect(height * 2).toEqual(heightB);
    });
    test('should increase height with line height increase', () => {
      const width = 100;
      const text = longTextB;
      const fontA = '100 4mm/4mm Arial';
      const height = calculateCellHeight(text, width, fontA, 1, 4);

      const fontB = '100 4mm/8mm Arial';
      const textB = longTextB;
      const heightB = calculateCellHeight(textB, width, fontB, 1, 8);
      expect(heightB).toBeGreaterThan(height);
    });

    test('should calculate cell height from text', () => {
      const cellHeight = calculateCellHeight(
        'abcdef',
        400,
        font,
        1,
        lineHeight
      );
      expect(cellHeight).toEqual(lineHeight * 1);
      expect(cellHeight % lineHeight).toEqual(0);
    });
    test('should calculate cell height from empty', () => {
      const cellHeight = calculateCellHeight('', 400, font, 1, lineHeight);
      expect(cellHeight).toEqual(lineHeight * 1);
      expect(cellHeight % lineHeight).toEqual(0);
    });
    test('should calculate cell height from long text', () => {
      const cellHeight = calculateCellHeight(
        `${longText}${longText}`,
        250,
        font,
        1,
        lineHeight
      );
      expect(cellHeight / lineHeight).toBeGreaterThan(1);
      expect(cellHeight % lineHeight).toEqual(0);
    });
    test('should calculate cell height from long text overflow', () => {
      const cellHeight = calculateCellHeight(
        `${longText}${longText}`,
        200,
        font,
        1,
        lineHeight
      );
      expect(cellHeight / lineHeight).toBeGreaterThan(1);
      expect(cellHeight % lineHeight).toEqual(0);
    });

    test('should strip out empty space at start and end', () => {
      const cellHeight = calculateCellHeight(
        `\ndef\n`,
        120,
        font,
        1,
        lineHeight
      );
      expect(cellHeight).toEqual(lineHeight * 1);
      expect(cellHeight % lineHeight).toEqual(0);
    });

    test('should strip out empty space at start and end with extra spaces', () => {
      const cellHeight = calculateCellHeight(
        `\ndef\n  \n`,
        120,
        font,
        1,
        lineHeight
      );
      expect(cellHeight).toEqual(lineHeight * 1);
      expect(cellHeight % lineHeight).toEqual(0);
    });
    test('should manage new line characters', () => {
      const cellHeight = calculateCellHeight(
        `abc\n\n\n def`,
        120,
        font,
        1,
        lineHeight
      );
      const lineCount = 4;
      expect(cellHeight).toEqual(lineHeight * lineCount);
      expect(cellHeight % lineHeight).toEqual(0);
    });

    test('should manage new line characters plus overflow', () => {
      const cellHeight = calculateCellHeight(
        `${longText}\n${longText}`,
        200,
        font,
        1,
        lineHeight
      );
      const lineCount = 4;
      expect(cellHeight).toEqual(lineHeight * lineCount);
      expect(cellHeight % lineHeight).toEqual(0);
    });

    test('should result in row height that is a multiple of line height', () => {
      const cellHeight = calculateCellHeight(longText, 60, font, 1, lineHeight);
      const lineCount = 6;
      expect(cellHeight).toEqual(lineHeight * lineCount);
      expect(cellHeight % lineHeight).toEqual(0);
      const cellHeightb = calculateCellHeight(
        longText,
        59,
        font,
        1,
        lineHeight
      );
      expect(cellHeightb).toEqual(lineHeight * lineCount);
      expect(cellHeightb % lineHeight).toEqual(0);
    });
    test('should work with RW example', () => {
      const text =
        '\nSome really long info about something that could be interesting and\nneeds to span lots of lines etc\n';
      const lineCount = 3;
      const width = 72;
      const pixelRatio = 3.76562;
      const cellPadding = 1;

      const cellHeight = calculateCellHeight(
        text,
        width - cellPadding,
        font,
        pixelRatio
      );
      const expectedHeight = (lineHeight * lineCount) / pixelRatio;
      expect(cellHeight).toEqual(expectedHeight);
      expect(expectedHeight % (lineHeight / pixelRatio)).toEqual(0);
    });
    describe('2 x pixel ratio', () => {
      const pixelRatio = 2;
      test('should calculate cell height from text', () => {
        const cellHeight = calculateCellHeight(
          'abcdef',
          400,
          font,
          pixelRatio,
          lineHeight
        );
        expect(cellHeight).toEqual((1 / pixelRatio) * lineHeight * 1);
        expect(cellHeight % (lineHeight / pixelRatio)).toEqual(0);
      });
      test('should calculate cell height from empty', () => {
        const cellHeight = calculateCellHeight(
          '',
          400,
          font,
          pixelRatio,
          lineHeight
        );
        expect(cellHeight).toEqual((1 / pixelRatio) * lineHeight * 1);
        expect(cellHeight % (lineHeight / pixelRatio)).toEqual(0);
      });
      test('should calculate cell height from long text', () => {
        const cellHeight = calculateCellHeight(
          `${longText}${longText}`,
          250,
          font,
          pixelRatio,
          lineHeight
        );
        const lineCount = 2;

        expect(cellHeight).toEqual((1 / pixelRatio) * lineHeight * lineCount);
        expect(cellHeight % (lineHeight / pixelRatio)).toEqual(0);
      });
    });
  });
});

describe('Calculate table column widths', () => {
  test('should fill table columns to table width using expand headers', () => {
    const tableWidth = 100;
    const headings = [
      {
        uid: 'h1',
        width: 20,
      },
      {
        uid: 'h2',
        width: 30,
        expand: true,
      },
      {
        uid: 'h2',
        width: 40,
      },
    ];
    const content = [['a', 'b', 'c']];
    const columnWidths = calculateTableColumnWidths(
      content,
      headings,
      tableWidth
    );
    expect(columnWidths).toEqual([20, 40, 40]);
    const calculatedTotalColumnsWidth = columnWidths.reduce(
      (acc, v) => acc + v,
      0
    );

    expect(calculatedTotalColumnsWidth).toEqual(tableWidth);
  });
  test('should fill table columns to table width with all headers if none set to expand', () => {
    const tableWidth = 100;
    const headings = [
      {
        uid: 'h1',
        width: 20,
      },
      {
        uid: 'h2',
        width: 37,
      },
      {
        uid: 'h2',
        width: 40,
      },
    ];
    const content = [['a', 'b', 'c']];
    const columnWidths = calculateTableColumnWidths(
      content,
      headings,
      tableWidth
    );
    expect(columnWidths).toEqual([21, 38, 41]);
    const calculatedTotalColumnsWidth = columnWidths.reduce(
      (acc, v) => acc + v,
      0
    );

    expect(calculatedTotalColumnsWidth).toEqual(tableWidth);
  });
});

describe('Process Element', () => {
  test('should process P', () => {
    const el = P('hello');
    const processedElement = processElement(el, DEFAULT_SETTINGS);
    expect(processedElement).toMatchSnapshot();
    expect(processedElement.text).toEqual('hello');
  });
  test('should not process Array', () => {
    // const el: ElementTypes[] = [P('hello'), P('world')];
    // TODO: Should not work
    // const processedElement = processElement(el, DEFAULT_SETTINGS);
    // expect(processedElement).toMatchSnapshot();
    // expect(processedElement[0].text).toEqual('hello');
  });
  test('should process table cell', () => {
    const el = TableCell('content', 'auto', 'auto', 'mixed', 2);
    const processedElement = processElement(el, DEFAULT_SETTINGS);
    expect(processedElement).toMatchSnapshot();
  });
  test('should process table cell with P', () => {
    const el = TableCell([P('content')], 'auto', 'auto', 'mixed', 2);
    const processedElement = processElement(el, DEFAULT_SETTINGS);
    expect(processedElement).toMatchSnapshot();
  });
});
