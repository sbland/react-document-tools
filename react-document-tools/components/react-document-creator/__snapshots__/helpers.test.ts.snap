// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process Element should process P 1`] = `
Object {
  "align": "left",
  "font": "100 4mm/4.5mm  Arial",
  "height": 4.5,
  "id": "MISSING_ID",
  "style": Object {},
  "text": "hello",
  "type": "p",
  "uid": 0,
}
`;

exports[`Process Element should process table cell 1`] = `
Object {
  "align": "left",
  "content": "content",
  "font": "100 4mm/4.5mm  Arial",
  "height": 4.5,
  "id": "MISSING_ID",
  "style": Object {},
  "type": "div",
  "uid": 1,
  "width": "auto",
}
`;

exports[`Process Element should process table cell with P 1`] = `
Object {
  "align": "left",
  "content": Array [
    Object {
      "align": "left",
      "font": "100 4mm/4.5mm  Arial",
      "height": 4.5,
      "id": "MISSING_ID",
      "style": Object {},
      "text": "content",
      "type": "p",
      "uid": 2,
    },
  ],
  "font": "100 4mm/4.5mm  Arial",
  "height": 4.5,
  "id": "MISSING_ID",
  "style": Object {},
  "type": "div",
  "uid": 3,
  "width": "auto",
}
`;
